Merge branch 'master' of https://github.com/ow2-proactive/scheduling into makeCleanupTimeoutConfigurable

tobwiens opened the pull request:
Make task termination cleanup time configurable

Problem: 
- Some tasks need time to cleanup after being killed (receiving SIGTERM in forked mode)
- This time can vary between use-cases

Solution:
- Allow do configure the cleanup waiting time through a system property

fviale commented on config/authentication/admin_user.cred:
```diff
@@ -1 +1 @@
-UlNBCjEwMjQKUlNBL0VDQi9QS0NTMVBhZGRpbmcKK6vW4qdV632zvbQUslGCi+guwX2bVZdC0pKq6RORLWWhQPL4ivRW+X0BpvYaaNINNzF/sAdSpkr2pUMMmBScJc5hXl9ECqrP0rWNimu49oiJ1EfEQV8XPAv2Drymj/okAS3r9bIXB/jv66vaAkrYpzm7ym2EvVqYR+UpQ6H+MN/lIpIMHSBOnleOExdlVjQNPXKnzmYld0zEQJx3B4vzYbCLNSJQUUbMzGFxrx452egcSRFM/+l0IFRmzm6tXGVo9WkW9LX+jU2GofAi5HP06hZIAMkq2GgDsuvdjLgWF/E8YfxtPhI+LxkA87Ka805ZKq+SBttk+85vc1uVrZrS8HFLaf6JezFnqK34MoH ...
\ No newline at end of file
+UlNBCjEwMjQKUlNBL0VDQi9QS0NTMVBhZGRpbmcKQPfY7B6TRKMu8+Xb09P5FTAqyYEHbNtrOdX8ljQ1PcsMCcSWQJZLukKE8pQJzjy9qwPNrNzqxan2aXDeFwUeES1Pcg4DlPEvAXu/3SgxndTKSvp0FCdLqPMtKBLPIkOU52x0yPeFGeWxBwykIrRxVLCypGsHjrER14b0JmDyn2b/LTUJU02lQcNBvISEIq0nOe0B04fZaou8k4gr02oO7cpMSx7K9ru7WDUG4KIYrk+V+22qDtNYdogvK6Wdixn3g+diuW2mfpPWW15f1ri5g0Tw9hE6vFwmJ3AwN9yOdBsS40iHTWCLDyzNHmUF6pMvGhECHPz3KmLBoAx+HFnOHfgcJ7yA9rV71lbItXy ...
```
You must exclude the credential files modifications from your pull request

fviale commented on scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/utils/TaskKiller.java:
```diff
@@ -67,16 +66,35 @@ private void interruptRepeatedly() {
 
+    private long getCleanupTimeMilliseconds() {
+        try {
+            String cleanupTimeString = System.getProperty("proactive.task.cleanup.time");
```
you should use a public static final name defined in RMNodeStarter

fviale commented on scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/utils/TaskKiller.java:
```diff
@@ -67,16 +66,35 @@ private void interruptRepeatedly() {
-            threadInterruptSleeper.sleep();
-        }
+        Sleeper cleanupTimeSleeper = getCleanupTimeSleeper();
+        cleanupTimeSleeper.sleep();
```
The problem with this implementation, is that the sleeper will wait for 10 or more seconds no matter if the threads actually finish quickly.

fviale commented on scheduler/scheduler-node/src/test/java/org/ow2/proactive/scheduler/task/utils/TaskKillerTest.java:
```diff
@@ -0,0 +1,111 @@
+
+public class TaskKillerTest {
+
+    private String taskKillerCleanupTimePropertyName = "proactive.task.cleanup.time";
```
same comment as above

tobwiens commented:
@fviale @paraita Can you have another look over the code. I fixed the following:
- RMNodeStarter is now the sole source of the property name
- Added test cases with garbage user input and negative numbers to insure that the node won't crash at the point when it kill a task. (That would we very hard to be traced back to the user input when the node was started).
- The waiting period is iterative, 100 milliseconds each iteration, which checks if the killing thread is still alive. So we don't waste time by blocking the node unnecessarily.

sonarqube-activeeon commented:
SonarQube analysis reported 1 issue

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [RMNodeStarter.java#L165](https://github.com/ow2-proactive/scheduling/blob/f9949d25fed4b0cf3e57349e23518f6462c7b57b/rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/RMNodeStarter.java#L165): Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

paraita commented on scheduler/scheduler-node/src/test/java/org/ow2/proactive/scheduler/task/utils/TaskKillerTest.java:
```diff
@@ -0,0 +1,196 @@
+        // Wait a millisecond for killing thread to start
+        waitOrFailTest(100);
+
+        assertThat("Task Killer must must interrupt once if kill() is called and then wait for the timeout which is set to 10 seconds.",
```
`must must` typo

fviale commented on scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/utils/TaskKiller.java:
```diff
@@ -67,15 +70,33 @@ private void interruptRepeatedly() {
+            }
+        } catch (NumberFormatException e) {
+            LOGGER.warn("proactive.task.cleanup.time: "
+                    + System.getProperty("proactive.task.cleanup.time")
```
please use the final static name here as well

tobwiens replied:
:+1:

fviale commented on scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/task/utils/TaskKiller.java:
```diff
@@ -36,15 +36,18 @@
+    private static final long CLEANUP_TIME_DEFAULT_SECONDS = 10;
     private static final long SLEEP_MILLISECONDS_DURING_THREAD_INTERRUPT = 10;
+    private static final long SLEEP_MILLISECONDS_BETWEEN_IS_THREAD_ALIVE_CHECK = 100;
+    private static final long MULTIPLIER_FROM_SECONDS_TO_ITERATION_WITH_CLEANUP_TIME_SLEEPER = 10;
```
This one can be computed = Math.floorDiv(1000, SLEEP_MILLISECONDS_BETWEEN_IS_THREAD_ALIVE_CHECK)

fviale replied:
Personnally, I would remove the declaration of this final variable, and simply write below:

long iterationsToHitWaitingTimeLimit = getCleanupTimeSeconds() * Math.floorDiv(1000, SLEEP_MILLISECONDS_BETWEEN_IS_THREAD_ALIVE_CHECK);

tobwiens replied:
I think that is taste. I like to have the name in the code so that the developer reads immediately what is happening instead of wondering why dividing through 1000.

fviale replied:
okay, but it still can be computed

sonarqube-activeeon commented:
SonarQube analysis reported 1 issue

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [RMNodeStarter.java#L165](https://github.com/ow2-proactive/scheduling/blob/4ee6768ba20d7c2e5891e0ab4fc2931fc189e6f5/rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/RMNodeStarter.java#L165): Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

tobwiens commented:
@fviale @paraita Implemented your comments. Squashed commits.

sonarqube-activeeon commented:
SonarQube analysis reported 1 issue

1. ![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) [RMNodeStarter.java#L165](https://github.com/ow2-proactive/scheduling/blob/79f6555d42063fcedd95baa9add7bd477e65f53e/rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/RMNodeStarter.java#L165): Reorder the modifiers to comply with the Java Language Specification. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AModifiersOrderCheck)

tobwiens commented:
@paraita @fviale Can you update your request changes so that we can merge this into todays release

tobwiens commented:
Retest this please!
